### PR TITLE
Karhunen loeve quadrature factory

### PR DESCRIPTION
**Pull request by regislebrun**
This class allows to compute approximate Karhunen-Loeve decomposition of covariance functions using a discrete approximation of the Fredholm equation.

Other improvements:
+ Added bounds computation to domain-related classes
+ Improved the draw() method of NumericalMathEvaluationImplementation: Now it is possible to use the 2D interface for 1D functions.

### CODE REVIEW

File: python/src/KarhunenLoeveQuadratureFactory_doc.i.in [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+----------
+domain : :class:`~openturns.Domain`
+    The domain on which the Fredholm equation is defined.
+experiment : :class:`~openturns.WeightedExperiment`
+    The algorithm used to discretize the integrale defining the Fredholm equation.
+basis : :class:`~openturns.Basis`
+    The basis from which the base functions are taken to define the finite space on which leave the approximate eigenfunctions.
+basisSize : int
+    The dimension of the finite approximation space.
+mustScale : logical
```
Comment (jschueller) [inline python/src/KarhunenLoeveQuadratureFactory_doc.i.in]: it's called bool

**Comment by jschueller** [comment]
+1
try to keep unrelated changes to tests as separate as possible

**Comment by sofianehaddad** [comment]
I think that 8ce75f2 should be in a separate PR

File: lib/src/Base/Geom/LevelSet.cxx [hunk 2]
```diff
@@ -125,6 +128,50 @@ void LevelSet::setLevel(const NumericalScalar level)
...
+  LinearNumericalMathFunction translate(NumericalPoint(1, level_), NumericalPoint(1), IdentityMatrix(1));
+  NumericalMathFunction equality(translate, function_);
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+    {
+      Matrix m(1, dimension);
+      m(0, i) = 1.0;
+      LinearNumericalMathFunction coordinate(NumericalPoint(dimension), NumericalPoint(1), m);
+      OptimizationProblem problem(coordinate, equality, NumericalMathFunction(), Interval());
+      problem.setMinimization(true);
+      Cobyla solver(problem);
```
Comment (dbarbier) [inline lib/src/Base/Geom/LevelSet.cxx]: Cobyla is hardcoded here, but this class should have a setOptimizationSolver to let users override this choice, shouldn't it?

**Comment by dbarbier** [comment]
+1

File: lib/src/Base/Geom/LevelSet.cxx (hunk 2 above)
Comment (regislebrun) [inline lib/src/Base/Geom/LevelSet.cxx]: @dbarbier There is no clean way to specify an optimization algorithm through ResourceMap. You provide either a string or an integer, and the mapping has to be done inside of the code in a static way (see eg GeneralizedLinearModelAlgorithm). The only way I found when working on the KrigingAlgorithm was to give access to the optimization problem to allow the user to solve it with its own solver, then to reinject the result through one of the constructors.

Do you have in mind an alternative way to do that? If not, I prefer to hardcode the algorithm waiting for a better management of complex types in ResourceMap.

**Comment by adutfoy** [comment]
+1

**Pull request merged**